dwarf-fortress: purge minor versions, add autoupdate, and fix audio

numinit opened the pull request:
Since DF saves are compatible within the same major version, we can simply keep the latest of each major version.
    
From Dwarf Therapist 42.1.7 onward, dfhack-generated memory maps work in place of Therapist's builtin memory maps. Allow this to happen.

Apparently we have to LD_PRELOAD the audio plugin (be it fmod or SDL2_mixer). Find an opportune time to do so during the startup script.

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

numinit committed:
dwarf-fortress: purge minor versions and add autoupdate

Since DF saves are compatible within the same major version, we can
simply keep the latest of each major version.

From Dwarf Therapist 42.1.7 onward, dfhack-generated memory maps work
in place of Therapist's builtin memory maps. Allow this to happen.

numinit commented on pkgs/games/dwarf-fortress/legends-browser/default.nix:
```diff
@@ -2,11 +2,11 @@
   jar = fetchurl {
     url = "https://github.com/robertjanetzko/LegendsBrowser/releases/download/${version}/legendsbrowser-${version}.jar";
-    sha256 = "05b4ksbl4481rh3ykfirbp6wvxhppcd5mvclhn9995gsrcaj8gx9";
+    sha256 = "sha256-jkv7InwaRn0K3VAa0LqkYpH6TnrT/tGYBtbvNGM6t98=";
```
hash

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/5077

bbigras commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 356659`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 20 packages built:</summary>
  <ul>
    <li>dwarf-fortress (dwarf-fortress-packages.dwarf-fortress ,dwarf-fortress-packages.dwarf-fortress_50_15)</li>
    <li>dwarf-fortress-packages.dwarf-fortress-full</li>
    <li>dwarf-fortress-packages.dwarf-fortress-original</li>
    <li>dwarf-fortress-packages.dwarf-fortress_0_43_05</li>
    <li>dwarf-fortress-packages.dwarf-fortress_0_44_05</li>
    <li>dwarf-fortress-packages.dwarf-fortress_0_44_09</li>
    <li>dwarf-fortress-packages.dwarf-fortress_0_44_10</li>
    <li>dwarf-fortress-packages.dwarf-fortress_0_44_11</li>
    <li>dwarf-fortress-packages.dwarf-fortress_0_44_12</li>
    <li>dwarf-fortress-packages.dwarf-fortress_0_47_01</li>
    <li>dwarf-fortress-packages.dwarf-fortress_0_47_02</li>
    <li>dwarf-fortress-packages.dwarf-fortress_0_47_04</li>
    <li>dwarf-fortress-packages.dwarf-fortress_0_47_05</li>
    <li>dwarf-fortress-packages.dwarf-fortress_50_10</li>
    <li>dwarf-fortress-packages.dwarf-fortress_50_11</li>
    <li>dwarf-fortress-packages.dwarf-fortress_50_12</li>
    <li>dwarf-fortress-packages.dwarf-fortress_50_13</li>
    <li>dwarf-fortress-packages.dwarf-fortress_50_14</li>
    <li>dwarf-therapist (dwarf-fortress-packages.dwarf-therapist)</li>
    <li>dwarf-fortress-packages.legends-browser</li>
  </ul>
</details>

numinit commented:
Thanks for the review. Should be easy to add the new Adventure Mode version in a separate PR.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/game-binary-dwarf-fortress-no-sound-on-nix-ld/61802/4

fpletz commented:
Why are all those versions necessary?

numinit commented:
Good question. It's mostly due to us historically just packaging them all, since you can't go backwards in version:

https://dwarffortresswiki.org/index.php/Save_compatibility

> Save compatibility is usually broken when a new major version is released, such as 50.01 or 0.40.01.

We could filter it down to major versions, I think there is still a case for that especially given differences between v49 and v50/v51.

numinit commented:
Alright, you've inspired me to do an autoupdate that sorts this out. :-)

numinit commented:
Autoupdate actually works now for this package set. I've tried a new strategy for this one that marks certain attributes as "expensive" (notably, output hashes and GH API queries) and recomputes them only if there's not a value for them in the previous lockfile.

I'm hoping to eventually adopt this strategy for other multi-version package sets like androidenv, since Google's repositories provide conflicting metadata that we may be able to resolve with some one-time expensive operations on the auto updater.

numinit committed:
dwarf-fortress: fix audio on v50

numinit committed:
release-notes: doc dwarf fortress improvements

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/game-binary-dwarf-fortress-no-sound-on-nix-ld/61802/5

numinit commented:
Audio is now fixed. You can now hear all the classic DF music in fortress mode, and on the menu.

numinit committed:
dwarf-fortress: expose dwarf-fortress-full and dfhack

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-in-distress/3604/105

numinit committed:
dwarf-fortress-packages.soundSense: fix initialization and drop JDK 8

We don't need the JDK 8 override anymore; soundsense works fine without
it if we use an --add-opens.

numinit commented:
I've tested the game and all the tools, we get sound in both Dwarf Fortress and Soundsense!

numinit commented:
This one was just removing lingering 0.43 versions.

tomberek commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 356659`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 1 package blacklisted:</summary>
  <ul>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 9 packages built:</summary>
  <ul>
    <li>dfhack (dwarf-fortress-full ,dwarf-fortress-packages.dwarf-fortress-full)</li>
    <li>dwarf-fortress (dwarf-fortress-packages.dwarf-fortress ,dwarf-fortress-packages.dwarf-fortress_51_10)</li>
    <li>dwarf-fortress-packages.dwarf-fortress-original</li>
    <li>dwarf-fortress-packages.dwarf-fortress_0_44_12</li>
    <li>dwarf-fortress-packages.dwarf-fortress_0_47_05</li>
    <li>dwarf-fortress-packages.dwarf-fortress_50_15</li>
    <li>dwarf-therapist (dwarf-fortress-packages.dwarf-therapist)</li>
    <li>dwarf-fortress-packages.legends-browser</li>
    <li>nixpkgs-manual</li>
  </ul>
</details>

tomberek reviewed (approved):
Runs w/sound as expected. Didn't have time to build a tavern.
